### PR TITLE
Ensure compliant init options API implementations.

### PR DESCRIPTION
**Pull request by hidmic**
Complying with https://github.com/ros2/rmw/pull/244.

### CODE REVIEW

**Comment by hidmic** [comment]
CI up to `rcl` and `test_rmw_implementation` against all RMW implementations:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11262/)

File: rmw_fastrtps_shared_cpp/src/rmw_init.cpp [hunk 1]
```diff
@@ -62,39 +68,47 @@ rmw_init_options_copy(
     return RMW_RET_INVALID_ARGUMENT;
   }
```
Comment (ivanpauno) [inline rmw_fastrtps_shared_cpp/src/rmw_init.cpp:69]: without line 66-69, this function could perfectly work with a non-zero initialized `dst` AFAIS

File: rmw_fastrtps_shared_cpp/src/rmw_init.cpp [hunk 2]
```diff
@@ -62,39 +68,47 @@ rmw_init_options_copy(
...
-  allocator.deallocate(init_options->enclave, allocator.state);
-  rmw_security_options_fini(&init_options->security_options, &allocator);
-  *init_options = rmw_get_zero_initialized_init_options();
-  return RMW_RET_OK;
+  rcutils_allocator_t * allocator = &init_options->allocator;
+  RCUTILS_CHECK_ALLOCATOR(allocator, return RMW_RET_INVALID_ARGUMENT);
+
+  rmw_ret_t ret = rmw_security_options_fini(&init_options->security_options, allocator);
+  if (RMW_RET_OK == ret) {
+    allocator->deallocate(init_options->enclave, allocator->state);
```
Comment (ivanpauno) [inline rmw_fastrtps_shared_cpp/src/rmw_init.cpp:108]: should this happen if security options failed to be deallocated?
maybe:
```cpp
allocator->deallocate(init_options->enclave, allocator->state);
rmw_ret_t ret = rmw_security_options_fini(&init_options->security_options, allocator);
if (RMW_RET_OK == ret) {
  *init_options = rmw_get_zero_initialized_init_options();
}
return ret;
```

File: rmw_fastrtps_shared_cpp/src/rmw_init.cpp (hunk 1 above)
Comment (hidmic) [reply rmw_fastrtps_shared_cpp/src/rmw_init.cpp:69]: I know, I prepared it for it. But it's lacking a finalization halfway through.

We can discuss guarantees later on and apply changes as needed.

File: rmw_fastrtps_shared_cpp/src/rmw_init.cpp (hunk 2 above)
Comment (hidmic) [reply rmw_fastrtps_shared_cpp/src/rmw_init.cpp:108]: So, this is something I wasn't clear about in the API contract. I'm actively trying to ensure a failed finalization doesn't affect state. Alternatively, we could take the finalize-as-much-as-you-can approach AND prepare finalization functions for multiple calls on partially finalized state. WDYT?

**Comment by hidmic** [comment]
One last sanity CI run after API contract updates:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11281/)

File: rmw_fastrtps_shared_cpp/src/rmw_init.cpp (hunk 2 above)
Comment (hidmic) [reply rmw_fastrtps_shared_cpp/src/rmw_init.cpp]: Alright, fini API contract has been updated.

**Pull request merged**